User management service

MostafaFouad0 opened the pull request:
# Overview
This PR introduces a microservice to handle user related requests.

# API endpoints implemented

| #   | Path                        | Method | Description                  |
| :-- | :-------------------------- | :----: | ---------------------------- |
| 1   | `/api/user/bio`             |  POST  | Change bio of a user         |
| 2   | `/api/user/profile`         |  GET   | get profile info of a user   |
| 3   | `/api/user/profile-pic`     |  POST  | change profile-pic           |
| 4   | `/api/user/cover-pic`       |  POST  | change cover-pic             |
| 5   | `/api/user/username`        |  POST  | change username              |
| 6   | `/api/user/follow`          |  POST  | follow a user                |
| 7   | `/api/user/follow`          | DELETE | unfollow a user              |
| 8   | `/api/user/followers/:id`   |  GET   | get all followers of a user  |
| 9   | `/api/user/followings/:id`  |  GET   | get all followings of a user |
| 10  | `/api/user/profile-pic/:id` |  GET   | get profile-pic of a user    |
| 11  | `/api/user/cover-pic/:id`   |  GET   | get cover-pic of a user      |

I improved the performance of several endpoints by leveraging RabbitMQ for message queuing and implementing asynchronous processing. This optimization reduces latency, enhances scalability, and ensures smoother handling of intensive tasks.

I will try to enhance the documentation of these endpoints in the `README.md` file.

MostafaFouad0 committed:
created the data model for the user and the followers tables

MostafaFouad0 committed:
updated the schema.

MostafaFouad0 committed:
implemented validator object using Joi

MostafaFouad0 committed:
implemented the validator middleware

MostafaFouad0 committed:
updated validator middlewarae

MostafaFouad0 committed:
generated primsa migration folder/files

MostafaFouad0 committed:
generated prisma migration folder/files

MostafaFouad0 committed:
updated the schema

MostafaFouad0 committed:
added dotenv library

MostafaFouad0 committed:
update the schema

MostafaFouad0 committed:
updated the schema

MostafaFouad0 committed:
implemented basic endpoints (getProfile, changeBio, changeUsername)

MostafaFouad0 committed:
implemented some basic endpoints (/avatar, /cover, /profile/:id)

MostafaFouad0 committed:
updated the endpoint names

MostafaFouad0 committed:
implemented endpoints (/profile-pic/id, /cover-pic/id) and changed the implementation of (/profile)

MostafaFouad0 committed:
implemented (getAllFollowers, getAllFollowings) endpoints

MostafaFouad0 committed:
fixed bug in server.js file

MostafaFouad0 committed:
reformated the code and the file structure
finished implementing all endpoints

MostafaFouad0 committed:
updated the schema of the follower table

MostafaFouad0 committed:
fixed a bug

MostafaFouad0 committed:
fixed a bug.

MostafaFouad0 committed:
removed some useless comments

MostafaFouad0 committed:
Updated the API documentation

Mohab96 commented on src/services/User Management/controllers/UserPicsControllers.js:
Please split these controllers into multiple files to enhance their readability a little bit and to help a reviewer to review or help yourself when something goes wrong to go debug the controller that is in question directly without reading the whole file. You can put them under the same directory, let's say `routes/controllers`, and inside `routes/` you can put the `Userroutes.js`. Additionally, you can add another directory `routes/validators`, and inside it you can do the same thing with the validators by splitting them everyone in a separate file.

You can take this as a reference: https://github.com/Mohab96/online-chess-game/tree/master/src/routes/auth

Mohab96 commented on src/services/User Management/controllers/FollowingManagementControllers.js:
Split them into multiple files.

Mohab96 commented on src/services/User Management/controllers/Usercontrollers.js:
Split them into multiple files.

Mohab96 commented on src/services/User Management/routes/Userroutes.js:
```diff
@@ -0,0 +1,38 @@
+const storage = multer.memoryStorage();
+const upload = multer({ storage: storage });
+
+router.post("/bio", changeBio);
```
For all these lines from 26 to 36, the lines where you assign an endpoint to a specific controller, you forgot to use the validators you implemented using `joi`.

MostafaFouad0 replied:
joi was for creating a new user, but this idea is outdated now since this functionality will be in the auth service

Mohab96 commented on src/services/User Management/validators/validator.js:
Split these validators to per endpoint validators instead of this universal schema variable and use every one of the new validators with the respective endpoint.
You can take this as reference: https://github.com/Mohab96/online-chess-game/tree/master/src/routes/auth/validators

Mohab96 commented on src/services/User Management/server.js:
Please use a random free port to run your service on so that you remove this responsibility off the shoulders of the `ServiceDiscovery` service of choosing a suitable port for it and also avoid any conflicts of running multiple microservices on the same port if you used the same way of hard coding the port in the code like this in other services.

You can take a look at this for reference: https://github.com/Mohab96/twitter-clone/blob/bd4faa4735d77c96fd4b3ce80aca847e89016a93/src/services/ServiceDiscovery/src/config/server.js#L13C1-L18C4

Note: I realized that some modifications will need to be made because the `ServiceDiscovery` service has no idea about which port every service will run on, so we might need to modify this to make the service run on random port by default iff a port was provided in the run command or in `env` file. But we will leave this for now and you just remove the hard coded port and we will deal with the rest of problems when we come to them.

Mohab96 commented on src/services/User Management/rmq/pushToRMQ.js:
I am not an expert in RabbitMQ, but I don't think opening a whole connection to just send this message and then closing it is the best choice, we do not do this also with the DB, we just open the connection and use them whenever we need them. From my humble opinion, I think this will add a significant overhead to the system (maybe I am exaggerating and it will not significant, but there will be overhead for sure). Consider opening a connection for pushing/consuming when you run the service for the first time.

Correct me if I am wrong because, as I said, I am not an expert in RabbitMQ.

MostafaFouad0 replied:
yes you are right, i totally forgot to update this section of the code.

Mohab96 commented on src/services/User Management/rmq/rmqConsumerServer.js:
In the future, consider using a logger for the whole service (e.g. `morgan`) or put some standards for the logging in your service and specify them in the service's `README.md` like when you will use `console.log`, when `console.error`, when `console.info`  and specify also where you will set the the logging level for this logging system. But as I said, this is future work so you can submit this PR if completely neglected this comment, it's totally fine.

Mohab96 reviewed (changes requested):
I left some comments, take a look at them and update your code if you see it needs to. Also consider putting some spaces in the future between lines in large files so that you separate your code "logically" into blocks, every one of these blocks does one thing.

Mohab96 commented on src/services/User Management/server.js:
```diff
@@ -0,0 +1,15 @@
+const consume = require("./rmq/rmqConsumerServer");
+const connectToDB = require("./prisma/connectToDB");
+const app = express();
+require("dotenv").config();
```
This should be the first or second line in the file so that if the database and RabbitMQ need to get anything from the `.env` file they will not be able to do this because this file is not configured yet.

Mohab96 commented:
Consider also adding `.env.example` in the future to state the variables that this service will need to run.

MostafaFouad0 committed:
made several changes

reformated the file strucure of the controllers

fixed the bug where every time the server want to send message to the RMQ it establishes new connection.

Mohab96 reviewed (approved):
Well done!
